pyro/ops/indexing.py

```diff
@@ -0,0 +1,95 @@
+from __future__ import absolute_import, division, print_function
+
+import torch
+
+
+def _is_batched(arg):
+    return isinstance(arg, torch.Tensor) and arg.dim()
+
+
+def vindex(tensor, args):
+    """
+    Vectorized advanced indexing with broadcasting semantics.
+
+    See https://www.numpy.org/neps/nep-0021-advanced-indexing.html
+
+    This assumes:
+
+    1. Each arg is either ``Ellipsis``, a ``slice``, or a batched
+        :class:`~torch.LongTensor` (i.e. with empty event shape).
+    2. If ``args[0] is not Ellipsis`` then ``tensor`` is not
+        batched, and its event dim is equal to ``len(args)``.
+    3. If ``args[0] is Ellipsis`` then ``tensor`` is batched and
+        its event dim is equal to ``len(args[1:])``.
+
+    :param torch.Tensor tensor: A tensor to be indexed.
+    :param tuple args: An index, as args to ``__getitem__``.
+    :returns: A nonstandard interpetation of ``tensor[args]``.
+    :rtype: torch.Tensor
+    """
+    if not isinstance(args, tuple):
+        args = (args,)
+    if not args:
+        return tensor
+
+    # Compute event dim before and after indexing.
+    if args[0] is Ellipsis:
+        args = args[1:]
+        if not args:
+            return tensor
+        old_event_dim = len(args)
+        args = (slice(None),) * (tensor.dim() - len(args)) + args
+    else:
+        args = args + (slice(None),) * (tensor.dim() - len(args))
+        old_event_dim = len(args)
+    assert len(args) == tensor.dim()
+    if any(a is Ellipsis for a in args):
+        raise NotImplementedError("Non-leading Ellipsis is not supported")
```

Comment:
This PR uses `Ellipsis` in a special way since we need to distinguish "batch dims" from "event dims" in the indexed tensor `x`. The meaning should coincide with integer indexing and with the named-vs-positional pattern in `Funsor` but I believe it differs from the NEP semantics which assumes all of x's dims are "event dims". Because we treat `Ellipsis` specially as a "batch shape" delimiter, we can't use it in the middle.

I'll add an explanation in the docstring.